test/test_nn.py

```diff
@@ -13356,6 +13338,77 @@ def test_grid_sample_nan_inf(self, device, dtype):
                                                      padding_mode=padding_mode, align_corners=False)
             self.assertEqual(sample, torch.zeros([1, 1, 1, 2], device=device, dtype=dtype))
 
+    @onlyOnCPUAndCUDA
+    def test_fractional_max_pool2d(self, device):
+        x = torch.randn(1, 2, 7, 7, requires_grad=True, device=device)
+        samples = x.new(1, 2, 2).uniform_()
+
+        def func(x):
+            return F.fractional_max_pool2d(
+                x, (2, 2), output_size=(3, 3), _random_samples=samples)
+
+        self.assertEqual(func(x).shape, (1, 2, 3, 3))
+        gradcheck(func, [x])
+        gradgradcheck(func, [x])
+
+        if self.device_type != 'cuda':
+            x = torch.randn(2, 7, 7, requires_grad=True, device=device)
+            samples = x.new(2, 2).uniform_()
+            self.assertEqual(func(x).shape, (2, 3, 3))
+            gradcheck(func, [x])
+            gradgradcheck(func, [x])
+
+        for kernel_size in [(), (1,)]:
+            with self.assertRaisesRegex(RuntimeError, "kernel_size must either"):
+                # Incorrect kernel_size
+                F.fractional_max_pool2d(x, kernel_size=kernel_size, output_size=(3, 3), _random_samples=samples)
+
+        err_large_msg = "too large relative to input "
+        err_out_size_msg = "output_size must either"
+        for output_size, msg in [((9, 3), err_large_msg + "height"),
+                                 ((3, 9), err_large_msg + "width"),
+                                 ((3,), err_out_size_msg),
+                                 ((), err_out_size_msg)]:
+            with self.assertRaisesRegex(RuntimeError, msg):
+                # Incorrect output_size
+                F.fractional_max_pool2d(x, (2, 2), output_size=output_size, _random_samples=samples)
+
+    @onlyOnCPUAndCUDA
+    def test_fractional_max_pool3d(self, device):
+        x = torch.randn(1, 2, 7, 7, 7, requires_grad=True, device=device)
+        samples = x.new(1, 2, 2).uniform_()
```

Comment:
Doesn't samples need one more element for the 3D case? The docs don't mention this args so I'm not sure what it should be.

Reply:
I think if you update the samples to ``samples = x.new(1, 2, 2, 2).uniform_()`` it will fix the errors.

Reply:
I think that makes sense but why does the CPU version pass in general. Though I think it fails due to this on the ASAN test.

Will try this. 
Thanks!

Reply:
`samples = x.new(1, 2, 2, 2).uniform_()` won't work as CUDA variant expects samples to be a 3-dim tensor.

https://github.com/pytorch/pytorch/blob/4ab0ef36a439a8d5f4b1203143df911b472a500a/aten/src/ATen/native/cuda/FractionalMaxPool3d.cu#L232-L245

Reply:
Correction, I looked more into random_samples and it's dimensions should always be 2D or 3D (batched). The dimensions are batch x channel x samples (1 sample per tensor dimension).

Reply:
Concretely, you need to change ``samples = x.new(1, 2, 2).uniform_()`` to ``samples = x.new(1, 2, 3).uniform_()``